modules/build/src/main/scala/scala/build/CrossSources.scala

```diff
@@ -142,19 +143,7 @@ object CrossSources {
       )
     }
 
-    val mainClassOpt = value {
-      inputs.mainClassElement
-        .collect {
-          case elem: Inputs.SingleElement =>
-            preprocessors.iterator
-              .flatMap(p => p.preprocess(elem).iterator)
-              .take(1).toList.headOption
-              .getOrElse(Right(Nil))
-              .map(_.flatMap(_.mainClassOpt.toSeq).headOption)
-        }
-        .sequence
-        .map(_.flatten)
-    }
+    val mainClassOpt = preprocessedSources.flatMap(_.mainClassOpt.toSeq).headOption
```

Comment:
It wasn't' necessary to run directive processing for mainClass because it was done before in `preprocessedSources`. 

I had a problem that the warning logging for the main class was printed twice.